### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by AndrewAR2**
Initial commit

### CODE REVIEW

**Comment by avive** [comment]
@AndrewAR2 - is this PR ready for code review?

**Comment by AndrewAR2** [comment]
The code is working on all platforms, however I have to add error handling.

**Comment by avive** [comment]
> The code is working on all platforms, however I have to add error handling.

Please do and let @noamnelke know when this PR is ready for code review.

**Comment by AndrewAR2** [comment]
I think this is ready for review.

File: test/test.go
Comment (noamnelke) [inline test/test.go]: This would be more useful (printing in hex):
```suggestion
		fmt.Printf("public key: %x\n", publicKey)
```

File: test/test.go
Comment (noamnelke) [inline test/test.go]: Print in hex:
```suggestion
		fmt.Printf("address: %x\n", address)
```

Also - this fails 😢

File: test/test.go
Comment (noamnelke) [inline test/test.go]: would be nice to use some non-zero value so we can see that encoding works (this should be printed, so we can easily confirm when running the test.

(not just on this line, all the values should be more arbitrary and not zeros)

File: test/test.go
Comment (noamnelke) [inline test/test.go]: If no device found it's hard to tell that the test didn't do anything (one could assume it passed).
```suggestion
	devices := ledger.GetDevices(0)
	if len(devices) == 0 {
		fmt.Printf("ERROR: no devices found\n")
	}
```

**Review by noamnelke — CHANGES_REQUESTED** [review]
Before finalizing this PR the following should be performed:
1. `go fmt ./...`
2. `go mod tidy`
3. `golint ./...` -> fix all issues (mostly capitalization and comments)

Some comments about the test file:
- The test requires a ledger device to be connected to the machine. We should implement an automated test using [Speculos](https://github.com/LedgerHQ/speculos).
- The test fails... The printed address output in the test does not match what the Ledger shows. Hard to tell if the public key is correct since the ledger doesn't display it.

File: test/test.go
Comment (noamnelke) [reply test/test.go]: D81CCA92-DF5F-44A7-A2D3-DEA635A7C95A

File: smesh.go
Comment (noamnelke) [inline smesh.go]: I may be nitpicking here, but the way to make this look nice in Go is like this:

```suggestion
const (
	P1_UNUSED     = 0x00
	P1_RETURN     = 0x01
	P1_DISPLAY    = 0x02
	P1_HAS_HEADER = 0x01
	P1_HAS_DATA   = 0x02
	P1_IS_LAST    = 0x04
)
```

This can be applied in many places in this PR, I just picked this one as an example.

Comment (noamnelke) [reply smesh.go]: Just to be clear, my version is after formatting.

File: test/test.go
Comment (AndrewAR2) [reply test/test.go]: Done!

File: test/test.go
Comment (AndrewAR2) [reply test/test.go]: Done!

File: hid_darwin.go
Comment (noamnelke) [inline hid_darwin.go]:
```suggestion
func GetDevices(productID int) []*Ledger {
```

**Review by noamnelke — CHANGES_REQUESTED** [review]
Please add a test to CI that makes sure the following don't make any changes (both currently fail):
1. `go fmt ./...`
2. `go mod tidy`
We have [tests like this](https://github.com/spacemeshos/go-spacemesh/blob/39fb64e0659d58e6bda689fe275586406ccc5ec3/Makefile#L150-L163), so you can use them as a reference.

Also add [a test](https://github.com/spacemeshos/go-spacemesh/blob/39fb64e0659d58e6bda689fe275586406ccc5ec3/Makefile#L165-L168) which checks that `golint ./...` and `go vet ./...` pass cleanly (`go vet` currently fails).

[This](https://github.com/spacemeshos/go-spacemesh/blob/39fb64e0659d58e6bda689fe275586406ccc5ec3/.github/workflows/ci.yml#L47-L68) is how we do it in Github Actions.

I didn't do a full review, I'll look at it again when these issues are resolved.

File: hid_darwin.go
Comment (AndrewAR2) [reply hid_darwin.go]: Fixed

**Comment by AndrewAR2** [comment]
Fixed!

File: ledger.go
Comment (noamnelke) [inline ledger.go:135]: Consider adding `Request Error 0x6E00: Spacemesh app not open on device` (happened to me and I had to look it up)

File: test/test.go
Comment (noamnelke) [inline test/test.go]: Consider moving binary data like this to a file. I'd add 3 files, one for each transaction type. Easier to load and makes the test easier to follow (less non-human-readable clutter).

File: test/test.go
Comment (noamnelke) [inline test/test.go]: This interface is not used

File: test/test.go
Comment (noamnelke) [inline test/test.go]: Why is the expected signature (for all transaction types) all zeros? I would expect to see a valid signature that can be validated against the actual transaction and public key.

File: test/test.go
Comment (noamnelke) [inline test/test.go]: This message is wrong - it should say something like "wrong signature".

This is just one example, but many "error" message say `[something]: OK\n` - this should be fixed.

File: test/test.go
Comment (noamnelke) [inline test/test.go]: This test looks like a copy-paste of the other test with some additional logic. I would consider extracting more logic pieces out of the test so we get code-reuse and easier-to-follow test code.

**Review by noamnelke — CHANGES_REQUESTED** [review]
General comments:

## Code comments
The code comments are messy and inconsistent. Go code comments are usually written as a natural English sentence, like:

```go
// WaitForClick waits until the user clicks the mouse button.
func WaitForClick(timeoutSecs int) {
```

It doesn't have to be limited to a single line/sentence, but the name of the function should be part fo the sentence and not repeated. The word after it should not be capitalized. Annotations like `@description` and `@param` are not needed. although parameters and return values should be explained if they're not obvious.

## Naming Constants
The convention/recommendation for naming constants in Go is the same as variables, i.e. camel case (starting with lower/upper case letter, depending on whether it needs to be exported). See this [SO answer](https://stackoverflow.com/a/22688926/1225467) for reference.

## Test
I'm glad to see we now have an automated test in CI using Speculos!

### Idiomatic Go Testing
Go has some great built-in testing infrastructure (using `go test`). We use [Testify](https://pkg.go.dev/github.com/stretchr/testify/assert) for assertions. I'm not sure it's worth the effort to convert the test, now that it's written, but it would surely be more idiomatic.

### Manual Test
The manual test should have better printed instructions. If no Ledger is detected, it's not unlocked or the Spacemesh app is not open - this should be printed to the user.

When the Ledger shows something that the user should verify (like the address path, etc.) the test should print this info so the user can actually verify it without checking the test code.

### Random Failure
I've had a failure the first time I ran the test. I haven't been able to reproduce it, so not sure we can properly investigate, but this was the output:
```
device: &{Path:IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/XHC1@14/XHC1@14000000/HS11@14600000/Nano S@14600000/Nano S@0/AppleUserUSBHostHIDDevice VendorID:11415 ProductID:4117 ReleaseNumber:0 UsagePage:65440 Usage:1 InterfaceNumber:0}
version: &{Major:1 Minor:0 Patch:0 Flags:0}
public key: &{872468e6347e9129a26a53392a71dcad90b7e8b1be346baf9908be9584ec3322 32851e50f638ef9d448302e7159ac2b452b9e4763679a26f6451457370bb00c3}
address: 872468e6347e9129a26a53392a71dcad90b7e8b1
show address: OK
Verify coin tx ERROR: writeHID error -1
```
After that point it was just stuck and nothing happened.

File: test/test.go
Comment (AndrewAR2) [reply test/test.go]: Fixed!

File: test/test.go
Comment (AndrewAR2) [reply test/test.go]: Fixed!

File: test/test.go
Comment (AndrewAR2) [reply test/test.go]: Fixed!

File: ledger.go
Comment (AndrewAR2) [reply ledger.go:135]: Fixed!

File: test/test.go
Comment (AndrewAR2) [reply test/test.go]: Done!

File: test/test.go
Comment (AndrewAR2) [reply test/test.go]: Done!

File: ledger_test.go
Comment (noamnelke) [inline ledger_test.go]:
```suggestion
	fmt.Printf("GetExtendedPublicKey test: Please confirm exporting the public key on your Ledger.\n")
```

Also, the Ledger shows the path that will be exported (`"44'/540'/0'/0/0'"`) so we should print that, too.

File: ledger_test.go
Comment (noamnelke) [inline ledger_test.go]:
```suggestion
	fmt.Printf("GetAddress test: Please confirm exporting the address on your Ledger.\n")
```

(print the path here, too)

File: ledger_test.go
Comment (noamnelke) [inline ledger_test.go]:
```suggestion
	fmt.Printf("ShowAddress test: Please make sure the following address agrees with your Ledger display.\n")
```

(print the path here, too)

File: ledger_test.go
Comment (noamnelke) [inline ledger_test.go]:
```suggestion
	path := "44'/540'/0'/0/0'"
	publicKey, err := device.GetExtendedPublicKey(StringToPath(path))
```

That way we can reuse the path string when printing and using with the api.

File: hid.go
Comment (noamnelke) [inline hid.go]:
```suggestion
	// LedgerUSBVendorID allows identifying USB devices made by Ledger.
```

File: hid.go
Comment (noamnelke) [inline hid.go]: No comments needed for interface methods (unless they help the reader understand something).

File: hid.go
Comment (noamnelke) [inline hid.go]:
```suggestion
	// ReadBuffMaxSize is the maximal number of bytes in the read buffer.
```

File: ledger.go
Comment (noamnelke) [inline ledger.go]:
```suggestion
	return response[:L-2], 0x9000
```

File: ledger.go
Comment (noamnelke) [inline ledger.go]:
```suggestion
		return nil, binary.BigEndian.Uint32(response[L-2:])
```
Your code is correct, but harder to read/interpret.

Note: you have to import `"encoding/binary"`.

File: ledger.go
Comment (noamnelke) [inline ledger.go]:
```suggestion
func (device *Ledger) send(cla, ins, p1, p2 byte, data []byte) ([]byte, error) {
```

File: ledger.go
Comment (noamnelke) [inline ledger.go]: Why do this in two steps?

File: tx-helper.go
Comment (noamnelke) [inline tx-helper.go]:
```suggestion
	binary.BigEndian.PutUint64(data, value)
```

File: tx-helper.go
Comment (noamnelke) [inline tx-helper.go]: Please handle the error here.

Also (you're converting to-from the same type):
```suggestion
	json.Unmarshal(byteValue, &data)
```

In general, you can unmarshal JSON directly to a struct in Go, but you might encounter some difficulties with hex strings and flattening the `data` field. While this function isn't pretty - maybe it's not worth the effort to improve it - I leave it up to you.

File: tx-helper.go
Comment (noamnelke) [inline tx-helper.go]: You could save many allocations by preallocating `tx` and using, for example:
```go
binary.BigEndian.PutUint64(tx[0:], txInfo.NetworkID)
tx[8] = txInfo.Type
binary.BigEndian.PutUint64(tx[9:], txInfo.Nonce)
copy(tx[17:], txInfo.To)
```
etc.

To simplify - you can still append `Data` and `PublicKey`.

File: utils.go
Comment (noamnelke) [inline utils.go]: This can also be improved

**Review by noamnelke — CHANGES_REQUESTED** [review]
I think we're getting close.

**Review by noamnelke — APPROVED** [review]
You've missed a few comments that were collapsed:
image


Once those are fixed - the PR can be merged.

Good job!! 👏🏻

**Comment by AndrewAR2** [comment]
Done! Thank you!

**Pull request merged**
